File: lib/tasks/update_stamp.rb
# frozen_string_literal: true

require "socket"
require "time"
require_relative "determinism"

module Tasks
  # Value object helpers for the per-record last-write stamp carried in JSONL.
  # Stamps stay as one sortable token on disk so the timestamp and device
  # tiebreaker cannot drift apart.
  module UpdateStamp
    VALUE_RE = /\A(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z)#([a-z0-9]+)\z/

    module_function

    def valid?(value)
      return false unless value.is_a?(String) && (match = VALUE_RE.match(value))

      Time.iso8601(match[1]).utc?
    rescue ArgumentError
      false
    end

    def key(value)
      return nil unless valid?(value)

      timestamp, device = value.split("#", 2)
      [timestamp, device]
    end

    def compare(left, right)
      left_key = key(left)
      right_key = key(right)
      return 0 if left_key.nil? && right_key.nil?
      return -1 if left_key.nil?
      return 1 if right_key.nil?

      left_key <=> right_key
    end

    def max(left, right)
      compare(left, right).negative? ? right : left
    end

    def format(time, device)
      normalized_device = slug(device)
      raise ArgumentError, "device slug is empty" if normalized_device.empty?
      raise ArgumentError, "clock must return a Time" unless time.respond_to?(:utc)

      "#{time.utc.strftime("%Y-%m-%dT%H:%M:%SZ")}##{normalized_device}"
    end

    # Hostnames commonly look like Marcus-MBP.local. The first alphanumeric
    # token is stable and short ("marcus") while still allowing an explicit
    # TASKS_DEVICE such as "home2" to disambiguate similar machines.
    def slug(value)
      value.to_s.downcase.split(".", 2).first.to_s.scan(/[a-z0-9]+/).first || "device"
    end

    # The hostname default goes through Determinism rather than straight to
    # Socket, because there are two hostname consumers — Config's host-context
    # selection and this device slug — and one pin has to cover both. With
    # TASKS_PIN_HOSTNAME unset, Determinism.hostname returns
    # `-> { Socket.gethostname }`, so the unpinned value is exactly what it was.
    # Resolved lazily: the default argument used to call Socket.gethostname on
    # every invocation even when TASKS_DEVICE made the answer irrelevant.
    def device(env: ENV, hostname: nil)
      override = env["TASKS_DEVICE"].to_s
      return slug(override) unless override.strip.empty?

      resolved = hostname.nil? ? Determinism.hostname(env: env).call : hostname
      slug(resolved.respond_to?(:call) ? resolved.call : resolved)
    end
  end
end

File: internal/updatestamp/updatestamp.go
// Package updatestamp is the per-record last-write stamp: one sortable token
// that keeps the timestamp and the device tiebreaker from ever drifting apart.
// It is the Go counterpart of lib/tasks/update_stamp.rb.
//
// The whole value object lives here — validation, the ordering key, comparison,
// and formatting — because the stamp is what decides last-write-wins when two
// devices edited the same record. A second spelling of "which stamp is newer"
// living in the merge path would be a second answer waiting to diverge.
package updatestamp

import (
	"regexp"
	"strings"
	"time"

	"github.com/marcus/tasks/internal/determinism"
)

// Slug reduces a hostname to its first alphanumeric token. Hostnames commonly
// look like Marcus-MBP.local; the first token is stable and short while still
// letting an explicit TASKS_DEVICE disambiguate similar machines.
func Slug(value string) string {
	head, _, _ := strings.Cut(strings.ToLower(value), ".")
	token := strings.Builder{}
	for index := 0; index < len(head); index++ {
		char := head[index]
		if (char >= 'a' && char <= 'z') || (char >= '0' && char <= '9') {
			token.WriteByte(char)
			continue
		}
		if token.Len() > 0 {
			break
		}
	}
	if token.Len() == 0 {
		return "device"
	}
	return token.String()
}

// Device is the device slug for this process: TASKS_DEVICE when it is set to
// anything but blank, otherwise the hostname — resolved through Determinism so
// that one pin covers both hostname consumers.
func Device(env determinism.Env) string {
	if override := env.Get(determinism.NameDevice); strings.TrimSpace(override) != "" {
		return Slug(override)
	}
	return Slug(determinism.Hostname(env)())
}

// valuePattern is UpdateStamp::VALUE_RE: one sortable token that keeps the
// timestamp and the device tiebreaker from drifting apart.
var valuePattern = regexp.MustCompile(`\A(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z)#([a-z0-9]+)\z`)

// Valid reports whether a stored `updated` value is a well-formed stamp: an
// RFC3339 UTC timestamp naming a real instant, then the device slug.
func Valid(value string) bool {
	match := valuePattern.FindStringSubmatch(value)
	if match == nil {
		return false
	}
	_, err := time.Parse("2006-01-02T15:04:05Z", match[1])
	return err == nil
}

// Key is the ordering key of a stamp: its timestamp half and its device half,
// or ok=false when the value is not a stamp at all. The two are compared in
// that order, which is why the stored spelling is one token — sorting the file
// by `updated` and sorting it by (timestamp, device) are the same sort.
func Key(value string) (timestamp, device string, ok bool) {
	if !Valid(value) {
		return "", "", false
	}
	timestamp, device, _ = strings.Cut(value, "#")
	return timestamp, device, true
}

// Compare orders two stamps the way UpdateStamp.compare does: -1, 0, or 1.
//
// An unparseable or absent stamp sorts BEFORE every real one, and two of them
// compare equal. That is the rule last-write-wins depends on: a record whose
// stamp a hand edit destroyed must never win against a record that carries a
// real one, and it must not oscillate when neither side has one.
func Compare(left, right string) int {
	leftTime, leftDevice, leftOK := Key(left)
	rightTime, rightDevice, rightOK := Key(right)
	switch {
	case !leftOK && !rightOK:
		return 0
	case !leftOK:
		return -1
	case !rightOK:
		return 1
	}
	if leftTime != rightTime {
		return strings.Compare(leftTime, rightTime)
	}
	return strings.Compare(leftDevice, rightDevice)
}

// Max is the later of two stamps, preferring the LEFT one on a tie — the same
// bias Ruby's `compare(left, right).negative? ? right : left` has, so a merge
// that finds two equal stamps keeps the value it already held.
func Max(left, right string) string {
	if Compare(left, right) < 0 {
		return right
	}
	return left
}

// Format renders the stamp a writer stores: the instant in UTC, then the device
// slug. Ruby additionally raises on an empty slug; Slug cannot return one (it
// falls back to "device"), so the guard is unreachable in both languages and is
// not reproduced as an error return.
func Format(instant time.Time, device string) string {
	return instant.UTC().Format("2006-01-02T15:04:05Z") + "#" + Slug(device)
}
